### PR TITLE
[WIN32SS][NTGDI] Avoid allocation of zero size in NtGdiGetGlyphIndicesW

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
In `NtGdiGetGlyphIndicesW` function, allocation of zero size had caused fatal failures.
JIRA issue: [CORE-12825](https://jira.reactos.org/browse/CORE-12825)
- Avoid allocation of zero size in `ExAllocatePoolWithTag` calls.
- Optimize for `cwc == 0`.

### CODE REVIEW

**Comment by katahiromz** [comment]
I cannot test because I don't have Word 2010.

File: win32ss/gdi/ntgdi/freetype.c [hunk 1]
```diff
@@ -6925,101 +6915,113 @@ NtGdiGetGlyphIndicesW(
...
-    {
-        DefChar = 0xffff;
-    }
-    else
-    {
-        FT_Face Face = FontGDI->SharedFace->Face;
-        if (FT_IS_SFNT(Face))
+        Buffer = ExAllocatePoolWithTag(PagedPool, cwc * sizeof(WORD), GDITAG_TEXT);
+        if (!Buffer)
+            return GDI_ERROR;
```
Comment (tkreuzer) [inline win32ss/gdi/ntgdi/freetype.c]: ERR("...")

File: win32ss/gdi/ntgdi/freetype.c [hunk 2]
```diff
@@ -6925,101 +6915,113 @@ NtGdiGetGlyphIndicesW(
     {
         return GDI_ERROR;
     }
-
     FontGDI = ObjToGDI(TextObj->Font, FONT);
     TEXTOBJ_UnlockText(TextObj);
 
-    Buffer = ExAllocatePoolWithTag(PagedPool, cwc*sizeof(WORD), GDITAG_TEXT);
-    if (!Buffer)
+    do
```
Comment (tkreuzer) [inline win32ss/gdi/ntgdi/freetype.c]: Please use "goto Exit" instead of this "loop".

File: win32ss/gdi/ntgdi/freetype.c [hunk 3]
```diff
@@ -6925,101 +6915,113 @@ NtGdiGetGlyphIndicesW(
...
+                Size = IntGetOutlineTextMetrics(FontGDI, 0, NULL);
+                if (!Size)
+                {
+                    cwc = GDI_ERROR;
+                    break;
+                }
+                potm = ExAllocatePoolWithTag(PagedPool, Size, GDITAG_TEXT);
+                if (!potm)
+                {
+                    cwc = GDI_ERROR;
```
Comment (tkreuzer) [inline win32ss/gdi/ntgdi/freetype.c]: This should be ```Status = STATUS_INSUFFICIENT_RESOURCES;```

File: win32ss/gdi/ntgdi/freetype.c (hunk 1 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: I used `DPRINT1` instread.

File: win32ss/gdi/ntgdi/freetype.c (hunk 2 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

File: win32ss/gdi/ntgdi/freetype.c (hunk 3 above)
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

**Comment by Doug-Lyons** [comment]
I have tested this with my Word 2010 and it works well for me. Thanks.

**Comment by katahiromz** [comment]
fabbd7c1c162dc4dec8a23f8470cafb8871d4675
https://build.reactos.org/builders/Build%20GCCLin_x86/builds/24602
https://build.reactos.org/builders/Test%20VBox/builds/11046

216129c701984f47712dd67778417e893ec5fe2d
https://build.reactos.org/builders/Build%20GCCLin_x86/builds/24603

**Comment by katahiromz** [comment]
`gdi32:GetGlyphIndices` has +1 failure to be investigated.

File: win32ss/gdi/ntgdi/freetype.c [hunk 4]
```diff
@@ -6923,29 +6914,32 @@ NtGdiGetGlyphIndicesW(
...
     {
+        DPRINT1("!TextObj\n");
         return GDI_ERROR;
     }
-
     FontGDI = ObjToGDI(TextObj->Font, FONT);
     TEXTOBJ_UnlockText(TextObj);
 
-    Buffer = ExAllocatePoolWithTag(PagedPool, cwc*sizeof(WORD), GDITAG_TEXT);
+    if (cwc <= 0)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

File: win32ss/gdi/ntgdi/freetype.c [hunk 5]
```diff
@@ -6879,15 +6879,18 @@ NtGdiGetGlyphIndicesW(
     OUTLINETEXTMETRICW *potm;
     INT i;
     WCHAR DefChar = 0xffff;
-    PWSTR Buffer = NULL;
+    PWSTR Buffer = NULL, Safepwc = NULL;
     ULONG Size, pwcSize;
-    PWSTR Safepwc = NULL;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

File: win32ss/gdi/ntgdi/freetype.c [hunk 6]
```diff
@@ -6923,29 +6914,32 @@ NtGdiGetGlyphIndicesW(
...
+        DPRINT1("!TextObj\n");
         return GDI_ERROR;
     }
-
     FontGDI = ObjToGDI(TextObj->Font, FONT);
     TEXTOBJ_UnlockText(TextObj);
 
-    Buffer = ExAllocatePoolWithTag(PagedPool, cwc*sizeof(WORD), GDITAG_TEXT);
+    if (cwc <= 0)
+        goto Exit;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

File: win32ss/gdi/ntgdi/freetype.c [hunk 7]
```diff
@@ -6954,11 +6948,18 @@ NtGdiGetGlyphIndicesW(
         else
         {
             Size = IntGetOutlineTextMetrics(FontGDI, 0, NULL);
+            if (!Size)
+            {
+                cwc = GDI_ERROR;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK.

File: win32ss/gdi/ntgdi/freetype.c [hunk 8]
```diff
@@ -7013,13 +7018,19 @@ NtGdiGetGlyphIndicesW(
     }
     _SEH2_END;
 
-ErrorRet:
-    ExFreePoolWithTag(Buffer, GDITAG_TEXT);
-    if (Safepwc != NULL)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply win32ss/gdi/ntgdi/freetype.c]: OK. Reverted.

**Comment by katahiromz** [comment]
4-failures
There are 4 failures in `gdi32:GetGlyphIndices`.

**Comment by katahiromz** [comment]
3-failure3-deleted
3 failures deleted. Retest on Word 2010, please!

**Comment by Doug-Lyons** [comment]
@katahiromz After your PR#1573 commit, these patches from Github no longer apply cleanly. Maybe you could squash and/or rebase these changes so that I can patch and test in Word 2010 again. Thanks.

**Pull request merged**